### PR TITLE
Account Protection: Fix TypeError in login_form_password_detection()

### PR DESCRIPTION
**Pull request by Initsogar**
Fixes p1745260735085449/1744645697.314949-slack-C080AF0NF0R

## Proposed changes:
* Add defensive coding to handle NULL values in `login_form_password_detection()`
* Add a unit test to verify NULL handling works correctly

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them? N/A (Backend only change)
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)? N/A

## Jetpack product discussion
p1745260735085449/1744645697.314949-slack-C080AF0NF0R

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:

The best way to verify this fix is through the unit tests, which directly test the NULL handling capability:

* Run unit tests and ensure it passes
```
cd projects/packages/account-protection
composer phpunit
```

This test specifically verifies that:
- When `login_form_password_detection()` receives a NULL user
- It returns NULL without attempting to call `user_requires_protection()`
- No fatal error occurs

In production, this prevents the TypeError when other authentication plugins return NULL during the filter chain.

### CODE REVIEW

File: projects/packages/account-protection/src/class-password-detection.php [hunk 1]
```diff
@@ -8,6 +8,7 @@
 namespace Automattic\Jetpack\Account_Protection;
 
 use Automattic\Jetpack\Assets\Logo as Jetpack_Logo;
+use WP_User;
```
Comment (tbradsha) [inline projects/packages/account-protection/src/class-password-detection.php]: Is this needed?

File: projects/packages/account-protection/src/class-password-detection.php [hunk 2]
```diff
@@ -41,13 +42,17 @@ public function __construct( ?Email_Service $email_service = null, ?Validation_S
...
-	 * @param string             $password The password.
+	 * @param \WP_User|\WP_Error|null $user The user or error object, or null.
+	 * @param string                  $password The password.
 	 *
-	 * @return \WP_User|\WP_Error The user object.
+	 * @return \WP_User|\WP_Error|null The user object, error object, or null.
 	 */
 	public function login_form_password_detection( $user, string $password ) {
-		if ( is_wp_error( $user ) || ! $this->user_requires_protection( $user, $password ) ) {
+		if ( is_wp_error( $user ) || $user === null || ! ( $user instanceof WP_User ) ) {
```
Comment (tbradsha) [inline projects/packages/account-protection/src/class-password-detection.php]: This check can be simplified:

```suggestion
		if ( is_wp_error( $user ) || ! ( $user instanceof WP_User ) ) {
```

File: projects/packages/account-protection/src/class-password-detection.php [hunk 3]
```diff
@@ -41,13 +42,17 @@ public function __construct( ?Email_Service $email_service = null, ?Validation_S
...
+			return $user;
+		}
+
+		if ( ! $this->user_requires_protection( $user, $password ) ) {
```
Comment (tbradsha) [inline projects/packages/account-protection/src/class-password-detection.php:56]: Technically this could be combined with the above condition, as PHP will stop evaluating at the first `true` in a chain of ORs, but keeping it separate is fine for readability.

File: projects/packages/account-protection/src/class-password-detection.php (hunk 2 above)
Comment (tbradsha) [reply projects/packages/account-protection/src/class-password-detection.php]: I'd suggest some adding a comment that documents why we need this extra check (third-party plugins have issues).

File: projects/packages/account-protection/tests/php/Password_Detection_Test.php [hunk 4]
```diff
@@ -504,4 +504,15 @@ public function test_render_content_shows_transient_error_if_set(): void {
 		$this->expectOutputRegex( '@' . $error['message'] . '@' );
 		$sut->render_content( $user, 'my_cool_token' );
 	}
+
+	/**
+	 * Tests that login_form_password_detection handles a NULL user gracefully without causing fatal errors.
```
Comment (tbradsha) [inline projects/packages/account-protection/tests/php/Password_Detection_Test.php:509]: ❤️

File: projects/packages/account-protection/src/class-password-detection.php (hunk 1 above)
Comment (Initsogar) [reply projects/packages/account-protection/src/class-password-detection.php]: No need, I just got rid of it and use it like `\WP_User` directly.

File: projects/packages/account-protection/src/class-password-detection.php (hunk 2 above)
Comment (Initsogar) [reply projects/packages/account-protection/src/class-password-detection.php]: Thanks, I just removed the explicit null check and added a comment :)

File: projects/packages/account-protection/src/class-password-detection.php (hunk 3 above)
Comment (Initsogar) [reply projects/packages/account-protection/src/class-password-detection.php:56]: Yes, I decided to kept it separate to improve readability. With the new comment on the type safety check above, it makes more sense to maintain these as distinct validation steps. Thanks for the suggestion!

**Comment by Initsogar** [comment]
Thanks for the review @tbradsha! I addressed your comments, could you please take another look? :)

**Review by tbradsha — APPROVED** [review]
